### PR TITLE
feat(flow): impl ScalarExpr&Scalar Function

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add impl for ScalarExpr&Scalar Functions

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Added impl code which eval Unary, Binary and Variadic Func

- Added utilty functions for `ScalarExpr` like get supporting(column being used in expression) column for a ScalarExpr.

- Most importantly, added utility functions for Temporal Filter, which can extract bounding condition including `now()` function and comparsion expression and express sliding windows etc.

- Summarize your change (**mandatory**)
useful methods for Scalar Expr, and `eval` method for Scalar function

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/flow/src/expr/scalar.rs [hunk 1]
```diff
@@ -59,3 +59,336 @@ pub enum ScalarExpr {
...
+                _ => unreachable!("Already checked"),
+            }
+        } else {
+            unsupported_err("None of the sides of the comparison is `now()`")
+        }
+    }
+}
+
+#[test]
+fn test_extract_bound() {
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:325]: Place under `tests` mod?

File: src/flow/src/expr/func.rs [hunk 2]
```diff
@@ -47,6 +44,62 @@ pub enum UnaryFunc {
...
+                let bool = if let Value::Boolean(bool) = arg {
+                    Ok(bool)
+                } else {
+                    TypeMismatchSnafu {
+                        expected: ConcreteDataType::boolean_datatype(),
+                        actual: arg.data_type(),
+                    }
+                    .fail()?
+                }?;
+                Ok(Value::from(!bool))
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:75]: 🤯 is it correct?

File: src/flow/src/expr/func.rs [hunk 3]
```diff
@@ -47,6 +44,62 @@ pub enum UnaryFunc {
...
+            }
+            Self::StepTimestamp => {
+                if let Value::DateTime(datetime) = arg {
+                    let datetime = DateTime::from(datetime.val() + 1);
+                    Ok(Value::from(datetime))
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:80]: I'm curious about what this function does, what's the purpose of `+ 1` ms?

File: src/flow/src/expr/func.rs [hunk 4]
```diff
@@ -96,8 +149,211 @@ pub enum BinaryFunc {
...
+        let left = expr1.eval(values)?;
+        let right = expr2.eval(values)?;
+        match self {
+            Self::Eq => Ok(Value::from(left == right)),
+            Self::NotEq => Ok(Value::from(left != right)),
+            Self::Lt => Ok(Value::from(left < right)),
+            Self::Lte => Ok(Value::from(left <= right)),
+            Self::Gt => Ok(Value::from(left > right)),
+            Self::Gte => Ok(Value::from(left >= right)),
+            Self::AddInt16 => Ok(add::<i16>(left, right)?),
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:168]:
```suggestion
            Self::Gte => Ok(Value::from(left >= right)),

            Self::AddInt16 => Ok(add::<i16>(left, right)?),
```

File: src/flow/src/expr/func.rs (hunk 3 above)
Comment (discord9) [reply src/flow/src/expr/func.rs:80]: It is useful when need to i.e. compare great then when default compare is not less than etc

> I'm curious about what this function does, what's the purpose of `+ 1` ms?

File: src/flow/src/expr/scalar.rs [hunk 5]
```diff
@@ -59,3 +59,336 @@ pub enum ScalarExpr {
...
+    /// each column referenced in `self`.
+    pub fn permute_map(&mut self, permutation: &BTreeMap<usize, usize>) {
+        self.visit_mut_post_nolimit(&mut |e| {
+            if let ScalarExpr::Column(old_i) = e {
+                *old_i = permutation[old_i];
+            }
+        });
+    }
+
+    pub fn support(&self) -> BTreeSet<usize> {
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:127]: doc required

File: src/flow/src/expr/scalar.rs [hunk 6]
```diff
@@ -59,3 +59,340 @@ pub enum ScalarExpr {
...
+        contains
+    }
+
+    /// extract lower or upper bound of `Now` for expr, where `lower bound <= expr < upper bound`
+    ///
+    /// returned bool indicates whether the bound is upper bound:
+    ///
+    /// false for lower bound, true for upper bound
+    /// TODO(discord9): allow simple transform like `now() + a < b` to `now() < b - a`
+    pub fn extract_bound(&self) -> Result<(Option<Self>, Option<Self>), String> {
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:272]: Can return `EvalError` as above?

File: src/flow/src/expr/scalar.rs [hunk 7]
```diff
@@ -59,3 +59,340 @@ pub enum ScalarExpr {
...
+                    BinaryFunc::Lte => BinaryFunc::Gte,
+                    BinaryFunc::Gt => BinaryFunc::Lt,
+                    BinaryFunc::Gte => BinaryFunc::Lte,
+                    _ => {
+                        return unsupported_err("The top level operator must be comparison");
+                    }
+                };
+            }
+            // TODO: support simple transform like `now() + a < b` to `now() < b - a`
+            if expr2.contains_temporal()
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:305]: 1. We can refactor this code to make it more readable
2. `contains_temporal` is a potentially heavy calculation, it is recommended to assign it to a variable

File: src/flow/src/expr/scalar.rs [hunk 8]
```diff
@@ -59,3 +59,341 @@ pub enum ScalarExpr {
...
+            ScalarExpr::CallUnary { func: _, expr } => f(expr),
+            ScalarExpr::CallBinary {
+                func: as_any,
+                expr1,
+                expr2,
+            } => {
+                f(expr1);
+                f(expr2);
+            }
+            ScalarExpr::CallVariadic { func: _, exprs } => {
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:206]: `func` can be omitted

File: src/flow/src/expr/scalar.rs [hunk 9]
```diff
@@ -59,3 +59,341 @@ pub enum ScalarExpr {
...
+        match self {
+            ScalarExpr::Column(_)
+            | ScalarExpr::Literal(_, _)
+            | ScalarExpr::CallUnmaterializable(_) => (),
+            ScalarExpr::CallUnary { func, expr } => f(expr),
+            ScalarExpr::CallBinary { func, expr1, expr2 } => {
+                f(expr1);
+                f(expr2);
+            }
+            ScalarExpr::CallVariadic { func, exprs } => {
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:240]: ditto

**Review by tisonkun — COMMENTED** [review]
nit: add a "flow" scope in the PR title (a.k.a. `feat(flow)`). It "shocks" me that we don't implement ScalarExpr for GreptimeDB, XD.

File: src/flow/src/expr/scalar.rs [hunk 10]
```diff
@@ -59,3 +61,340 @@ pub enum ScalarExpr {
...
+            {
+                std::mem::swap(&mut expr1, &mut expr2);
+                func = BinaryFunc::reverse_compare(&func)?;
+            }
+            // TODO: support simple transform like `now() + a < b` to `now() < b - a`
+            if expr_2_contains_now
+                || *expr1 != ScalarExpr::CallUnmaterializable(UnmaterializableFunc::Now)
+            {
+                return unsupported_err("None of the sides of the comparison is `now()`");
+            }
```
Comment (zhongzc) [inline src/flow/src/expr/scalar.rs:299]:
```suggestion
            let expr1_is_now = *expr1 == ScalarExpr::CallUnmaterializable(UnmaterializableFunc::Now);
            let expr2_is_now = *expr2 == ScalarExpr::CallUnmaterializable(UnmaterializableFunc::Now);
            if !(expr1_is_now ^ expr2_is_now) {
                return unsupported_err("one side of the comparison must be `now()`");
            }
            
            if expr2_is_now {
                std::mem::swap(&mut expr1, &mut expr2);
                func = BinaryFunc::reverse_compare(&func)?;
            }
```

**Review by zhongzc — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**